### PR TITLE
3 - the filter ecosystem evolved

### LINKED ISSUE #37: Create a persistent filter that can be applied to a participant set
We want to be able to save a filter over the information of a participant, so we can follow up on this at a later point in time (e.g. if participants were added post-filtercreation that match the filter) and the participant can save the search in their dashboard.

This includes

- [ ] creating a custom filter for the location and time info
- [ ] creating a filter that is generated from the participant info (it might be configurable, but it should not require a config, maybe as a later issue and for now use defaults). This could mean that a text property could allow open text search, bool fields have yes/no/none, choices can be multiselected and ordered choices can be range selected etc. Some of that functionality could be reused from `django-filter`. 

#### Open Question
Do we allow more intricate logic to compose the filters (and, or...)? Facebooks search graph was an experiment that allowed this but I guess most of the population never used the features that the developers created.... ;)

### LINKED ISSUE #61: Make the location filter editable in the UI and support several locations as a property
Connected to the discussion on matching modi in #45

### LINKED ISSUE #170: Link the correct filter view in the map
As a user I want to soom. into the map, when I click on detail, I should be linked to a list of people at that place

### LINKED ISSUE #176: Optional name for filters
As a user with many filters I can easily be lost with all my filters but as a users with only few filters I don't want to understand what a name of a filter is.

Add a standard name or empty text field to the list of filters which can be assigned by power users but doesn't have to by standard user

### LINKED ISSUE #188: Allow users to see who they contacted and when and share their email via our plattform

### LINKED ISSUE #195: Particiapnts are exchanged on the map
Actually, B is A and A is B.

### PR DESCRIPTION
**Pull request by feeds**
Fixes #195 
Fixes #170 
Fixes #61
Fixes #176
Closes #37
Closes #188

### CODE REVIEW

**Review by Baschdl — COMMENTED** [review]
Should I be able to see the filter page as a not-logged-in user? I can't see the details of a user but I can determine them with the filters.
Does the filter page open up on a separate page on purpose?

**Comment by Baschdl** [comment]
I am able to try to send mails as not-logged-in user which luckily fails. But if it's intended that a user can see the filter page, maybe we should not show the contact option.
image

**Comment by Baschdl** [comment]
I can contact new matches and get a positive feedback even if there aren't any new matches.
image
It even appears as a contact:
image

File: backend/apps/matching/templates/participant/participant_dashboard.html [hunk 1]
```diff
@@ -34,6 +35,53 @@
...
+                            {% if not_responded_to_n_requests != 0%}
+                            <span class="badge badge-warning">{{ not_responded_to_n_requests    }}</span>
+                            {%  endif %}
+                        </a>
+    </div>
+    <div class="col-md-4">
+                        <a role="button" class="btn btn-outline-primary btn-sm btn-block"
+                           href="{% url 'matches-requests-from-me' %}">
+                            <i class="fa fa-mail-forward" aria-hidden="true"></i>
+                            {% blocktrans %}Manage outgoing contact requests{% endblocktrans %}
```
Comment (Baschdl) [inline backend/apps/matching/templates/participant/participant_dashboard.html]:
```suggestion
                            {% blocktrans %}See outgoing contact requests{% endblocktrans %}
```

File: backend/locale/de/LC_MESSAGES/django.po [hunk 2]
```diff
@@ -177,12 +342,29 @@ msgstr ""
...
+msgstr "Ihr Dashboard"
+
+msgid "Start a new search for "
+msgstr "Erstelle eine neue Suche nach"
+
+msgid "Respond to incoming contact requests"
+msgstr "Auf eingehende Anfragen antworten"
+
+msgid "Manage outgoing contact requests"
+msgstr "Ausgehende Anfragen verwalten"
```
Comment (Baschdl) [inline backend/locale/de/LC_MESSAGES/django.po]:
```suggestion
msgid "View outgoing contact requests"
msgstr "Ausgehende Anfragen ansehen"
```

File: backend/apps/matching/configuration/properties.py [hunk 3]
```diff
@@ -205,6 +220,18 @@ def generate_random_assignment(self, rs=None):
...
+    def get_filters(self) -> List:
+        return [
+            [
+                {
+                    "lookup_exp": "exact",
+                    "label": _("is exactly"),
+                    "model_field": models.NullBooleanField(blank=True, null=True),
+                }
+            ]
+            for i in range(len(self.choices))
```
Comment (Baschdl) [inline backend/apps/matching/configuration/properties.py:326]: Is this the modern equivalent to a blank line?

**Review by bjrne — CHANGES_REQUESTED** [review]
nur relevant für Maltes UI: "Contact all selected" Könnte "Contact X people" sein, einfach ein bisschen mehr feedback für den user

- [x] As a logged in type A user: Wenn ich eine Mail an 2 leute verschicken will kriege ich einen IntegrityError at matching/A/

nur relevant für maltes UI: "Contact selected"-Button Öffnet ein email fenster wenn man niemanden ausgewählt hat
nur relevant für maltes UI: Checkbox in table header markiert nicht alle filter results (i guess das ist die idee)

Was ist "Contact enthält" als label im Filter? Email?

File: backend/apps/matching/models/participant_info_location.py [hunk 4]
```diff
@@ -13,10 +13,11 @@
 
 
 class RadiusChoices(models.IntegerChoices):
+    ONLY_HERE = 0, _("0 km")
     LESSTEN = 10, _("<10 km")
     LESSTWENTY = 20, _("<20 km")
-    LESSFOURTY = 30, _("<40 km")
-    MOREFOURTY = 40, _(">40 km")
+    LESSFOURTY = 40, _("<40 km")
+    MOREFOURTY = 50, _("<50 km")
```
Comment (Baschdl) [inline backend/apps/matching/models/participant_info_location.py]:
```suggestion
    LESSFIFTY = 50, _("<50 km")
```

File: backend/apps/matching/models/participant_info_location.py [hunk 5]
```diff
@@ -31,6 +32,8 @@ class AbstractParticipantInfoLocation(models.Model):
     Only for Germany and Austria.
     """
 
+    MAX_RADIUS = 100  # km - limits the search APIs radius
```
Comment (Baschdl) [inline backend/apps/matching/models/participant_info_location.py]: That's 2.2 million people in Meck-Pom and 128.9 million people in Mega-Berlin.

File: backend/apps/matching/tables/filters.py [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+    class FilterTable(tables.Table):
+        registration_date = tables.Column(accessor="registration_date", verbose_name=_("Created"))
+        filter_name = tables.TemplateColumn(
+            template_name="filters/table_name_col.html",
+            extra_context={"p_type": p_type},
+            verbose_name=_("Name"),
+        )
+        matches = tables.Column(empty_values=(), verbose_name=_("New matches"))
+        new_matches_you_did_not_contact = tables.TemplateColumn(
+            template_name="filters/contact_new_match_col.html", verbose_name="", empty_values=()
```
Comment (Baschdl) [inline backend/apps/matching/tables/filters.py]:
```suggestion
            template_name="filters/contact_new_match_column.html", verbose_name="", empty_values=()
```

File: backend/apps/matching/templates/filters/filter_list.html [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+{% load static %}
+{% load i18n %}
+
+{% block title %}
+{% blocktrans %}Your searches{% endblocktrans %}
+{% endblock %}
+
+{% block header %}
+<link rel="stylesheet" href="{% static 'css/form.css' %}">
+<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css">
```
Comment (Baschdl) [inline backend/apps/matching/templates/filters/filter_list.html]: No no. @maltezacharias

File: backend/apps/matching/templates/map/map.html [hunk 8]
```diff
@@ -17,8 +17,8 @@
       mapViewContainerId: 'mapContainer',
       aURL: "{% url 'participant_JSON' 'A' %}",
       bURL: "{% url 'participant_JSON' 'B' %}",
-      aListURL: "{% url 'participant_list' 'A' 'COUNTRYCODE' 'PLZ' 0 %}",
-      bListURL: "{% url 'participant_list' 'B' 'COUNTRYCODE' 'PLZ' 0 %}",
+      aListURL: "{% url 'participant_list' 'A'%}?location_country_code=COUNTRYCODE&location_zipcode=PLZ&location_distance=0",
+      bListURL: "{% url 'participant_list' 'B'%}?location_country_code=COUNTRYCODE&location_zipcode=PLZ&location_distance=0",
```
Comment (Baschdl) [inline backend/apps/matching/templates/map/map.html:21]: Is this really the django way to do it?

File: backend/apps/matching/templates/matches/matches_from_others_list.html [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+{% load static %}
+{% load i18n %}
+
+{% block title %}
+{% blocktrans %}Contact Requests{% endblocktrans %}
+{% endblock %}
+
+{% block header %}
+<link rel="stylesheet" href="{% static 'css/form.css' %}">
+<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css">
```
Comment (Baschdl) [inline backend/apps/matching/templates/matches/matches_from_others_list.html]: no no no

File: backend/apps/matching/templates/matches/matches_to_others_list.html [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+{% load static %}
+{% load i18n %}
+
+{% block title %}
+{% blocktrans %}Your searches{% endblocktrans %}
+{% endblock %}
+
+{% block header %}
+<link rel="stylesheet" href="{% static 'css/form.css' %}">
+<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css">
```
Comment (Baschdl) [inline backend/apps/matching/templates/matches/matches_to_others_list.html]: no

File: backend/locale/de/LC_MESSAGES/django.po [hunk 11]
```diff
@@ -358,6 +569,12 @@ msgid "Please validate your email by clicking this:"
...
 "Bitte validieren Sie Ihre E-Mail indem Sie den folgenden Link anklicken:"
 
+msgid "Disapprove"
+msgstr "Ablehnen"
+
+msgid "Approve"
+msgstr "Freischalten"
+
 msgid "Approvals"
 msgstr "Freischalten"
```
Comment (Baschdl) [inline backend/locale/de/LC_MESSAGES/django.po:566]:
```suggestion
msgstr "Freischaltungen"
```

**Review by Baschdl — CHANGES_REQUESTED** [review]

**Comment by Baschdl** [comment]
Did the "Share mail" with institutions-button come from user feedback (a.k.a. is this really a useful feature)? Does this mean that a user only gets the mails from institutions after clicking this button?

**Comment by feeds** [comment]
idea was first to somehow track wether the first contact was successful to give e. g. institutions a better overview over who they contacted (i think this was user feedback). So we thought of linking this 'share email field' int he email. 

but on second thought idea could be to generate a direct link under the message to the requested user that opens the response form that already exists (when you klick on 'message') Then we can register this in our system and afterwarda email contact is without us

**Comment by feeds** [comment]
with regards to permissions: i did not waste a thought on that, as I said before, until all other permissions are cleae and the relevant decorators are available

**Comment by feeds** [comment]
> I can contact new matches and get a positive feedback even if there aren't any new matches.
> image
> It even appears as a contact:
> image
> 

okay, thats not intended behaviour. But it stems from the fact that new matches are people that match and we did not contact at all (but it might be that we did not contaxr them via this exacr filter)

is a todo to copy that behaviour to the contact field and disable it if there are no new matches

**Comment by feeds** [comment]
for seeing the filter list of oneselfb (send email fails) 

i wanted to get some input here - i think it would be good to ba able for B to see the filter with which A contacted it but not search with it (=> would need a static view of the persistent filter)

File: backend/apps/matching/models/participant_info_location.py (hunk 5 above)
Comment (feeds) [reply backend/apps/matching/models/participant_info_location.py]: thats 5 people in Schleswig-Holstein

Comment (feeds) [reply backend/apps/matching/models/participant_info_location.py]: but i would be ok if we make it a config ;)

File: backend/apps/matching/templates/map/map.html (hunk 8 above)
Comment (feeds) [reply backend/apps/matching/templates/map/map.html:21]: I tried googleing it already and apparently it is.
other way is passing a dict with get params to urlencode.

**Comment by maltezacharias** [comment]
Will review tonight, please don't merge until tomorrow

File: backend/apps/matching/configuration/properties.py (hunk 3 above)
Comment (feeds) [reply backend/apps/matching/configuration/properties.py:326]: das ist der modern equivalent zu nicht required

**Comment by feeds** [comment]
Done
- as A view search with only when of type A, for B see only the criteria
- contact the right people when clicking on contact new
- max radius in participant config
- remove link tags
- fix the "contact new" button to only work when there are actually new people

Todo
- (translate)

Other people:
- js for the select all checkbox (=> malte will do a new ui component anyways)
- nice ui for viewing the filter criteria

Later
- permissions

**Pull request merged**